File: aurora-core/src/main/java/net/gree/aurora/domain/datasource/DataSourceIdResolver.java
package net.gree.aurora.domain.datasource;

/**
 * ヒントから{@link DataSourceId}を解決するためのインターフェイス
 *
 * @param <T> ヒントの型
 */
public interface DataSourceIdResolver<T> {

    /**
     * 関数にヒントを適用して戻り値を得る。
     *
     * @param hint ヒント
     * @return {@link DataSourceId}
     */
    DataSourceId apply(T hint);

}

File: aurora-scala/src/main/scala/net/gree/aurora/scala/domain/datasource/DataSourceIdResolver.scala
package net.gree.aurora.scala.domain.datasource

import scala.language.implicitConversions
import net.gree.aurora.domain.datasource.{DataSourceIdResolver => JDataSourceIdResolver}

/**
 * [[net.gree.aurora.domain.datasource.DataSourceIdResolver]]のScala版。
 *
 * @tparam T ヒントの型
 */
trait DataSourceIdResolver[T] extends (Option[T] => DataSourceId) {

  def apply(value: Option[T] = None): DataSourceId

}

/**
 * コンパニオンオブジェクト。
 */
object DataSourceIdResolver {

  /**
   * ScalaオブジェクトをJavaオブジェクトに変換する。
   *
   * @param dataSourceIdResolver [[net.gree.aurora.scala.domain.datasource.DataSourceIdResolver]]
   * @tparam T ヒントの型
   * @return [[net.gree.aurora.domain.datasource.DataSourceIdResolver]]
   */
  implicit def toJava[T](dataSourceIdResolver: DataSourceIdResolver[T]) = {
    new JDataSourceIdResolver[T] {
      def apply(hint: T) = dataSourceIdResolver(Option(hint))
    }
  }

}
